File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.cc
#include <iostream>
#include <queue>

int dx[8] = {-2,-1,1,2,2,1,-1,-2};
int dy[8] = {1,2,2,1,-1,-2,-2,-1};

using namespace std;

void bfs(int start, int end, int visited[301][301], int goalx, int goaly, int len, bool arr[301][301]){
  queue<pair<int, int>>q;
  q.push({start, end});
  visited[start][end]=1;
  arr[start][end]=true;
  while(!q.empty()){
    int front=q.front().first;
    int back=q.front().second;
    q.pop();
    // cout<<front<<" "<<back<<" "<<visited[front][back]<<endl;

    if(front==goalx && back==goaly){
      cout<<visited[front][back]-1<<endl;
      return ;
    }
    for(int i=0;i<8;i++){
      int nx=front+dx[i];
      int ny=back+dy[i];
      if(nx>=0 && ny>=0 && nx<len && ny<len && arr[nx][ny]==false  && visited[nx][ny]==0){
        visited[nx][ny]=visited[front][back]+1;
        q.push({nx,ny});
        arr[nx][ny]=true;
      }
    }

    

  }

}
int main(){
  int t;
  cin>>t;
  for(int i=0;i<t;i++){
    int visited[301][301];
    bool arr[301][301];
    int len;
    cin>>len;
    for(int i=0;i<len;i++){
      for(int j=0;j<len;j++){
        visited[i][j]=0;
      }
    }
    for(int i=0;i<len;i++){
      for(int j=0;j<len;j++){
        arr[i][j]=false;
      }
    }
    int px, py;
    cin>>px>>py;
    int gx,gy;
    cin>>gx>>gy;
    if(px==gx && py==gy){
      cout<<'0'<<endl;
      continue;
    }
    bfs(px, py, visited, gx, gy, len, arr);    
  }
  

}

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;
public class Main {
	public static int[] dx = {2,2,-2,-2,1,-1,1,-1};
	public static int[] dy = {1,-1,1,-1,2,2,-2,-2};
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int num = Integer.parseInt(br.readLine());
		for(int i=0;i<num;i++) {
			Queue<Pair> q = new LinkedList<Pair>();
			int size = Integer.parseInt(br.readLine());
			int[][] arr = new int[size][size];
			boolean[][] visited = new boolean[size][size];
			String[] start = br.readLine().split(" ");
			int startX = Integer.parseInt(start[0]);
			int startY = Integer.parseInt(start[1]);
//			arr[startX][startY] = 1;
			String[] end = br.readLine().split(" ");
			int endX = Integer.parseInt(end[0]);
			int endY = Integer.parseInt(end[1]);
			q.add(new Pair(startX, startY));
			visited[startX][startY] = true;
			while(!q.isEmpty()) {
				Pair p = q.poll();
				int x = p.x;
				int y = p.y;
				if(x == endX && y == endY) break;
				for(int j=0;j<8;j++) {
					int nx = x+dx[j];
					int ny = y+dy[j];
					if(nx>=0 && ny>=0 && nx<size && ny<size && !visited[nx][ny] && arr[nx][ny] == 0) {
						arr[nx][ny] = arr[x][y]+1;
						visited[nx][ny] = true;
						q.add(new Pair(nx,ny));
					}
				}
			}
			System.out.println(arr[endX][endY]);
		}
	}
	
	public static class Pair{
		int x;
		int y;
		Pair(int x, int y) {
			this.x= x;
			this.y = y;
		}
	}
}
